internal/pkg/client/cache/dir.go

```diff
@@ -22,69 +22,251 @@ const (
+		return false
+	}
+
+	return true
+}
+
+// Clean wipes all files in the cache directory, will return a error if one occurs
+// Since renamed Destroy() but kept for backward compatibility
+func (c *SingularityCache) Clean() error {
+	return c.Destroy()
+}
+
+func isDefaultBasedir(basedir string) (bool, error) {
 	usr, err := user.Current()
 	if err != nil {
-		sylog.Fatalf("Couldn't determine user home directory: %v", err)
+		return false, fmt.Errorf("failed to get user: %s", err)
 	}
 
-	if d := os.Getenv(DirEnv); d != "" {
-		root = d
-	} else {
-		root = path.Join(usr.HomeDir, RootDefault)
+	if basedir == path.Join(usr.HomeDir, BasedirDefault) {
+		return true, nil
 	}
 
-	if err := initCacheDir(root); err != nil {
-		sylog.Fatalf("Unable to initialize caching directory: %v", err)
+	return false, nil
+}
+
+// Figure out where the cache directory is.
+func getCacheBasedir() (string, error) {
+	// If the user defined the special environment variable, we use its value
+	usr, err := user.Current()
+	if err != nil {
+		return "", fmt.Errorf("couldn't determine user home directory: %s", err)
+	}
+
+	// Assuming the user set the environment variable
+	basedir := os.Getenv(DirEnv)
+	if basedir == "" {
+		basedir = path.Join(usr.HomeDir, BasedirDefault)
 	}
+
+	return basedir, nil
 }
 
-func updateCacheSubdir(subdir string) string {
-	updateCacheRoot()
+// getCacheRoot figures out what the root directory is
+func getCacheRoot(basedir string) string {
+	return path.Join(basedir, RootDefault)
+}
+
+// updateCacheSubdir update/create a sub-cache (directory) within the cache,
+// for example, the 'shub' cache.
+func updateCacheSubdir(c *SingularityCache, subdir string) (string, error) {
+	// This function may act on an cache object that is not fully initialized
+	// so it is not a method on a SingularityCache but rather an independent
+	// function.
+	// Because the cache object may not be initialized, we do NOT check its validity
+	if c == nil {
+		return "", fmt.Errorf("invalid cache handle")
+	}
+
+	// If the subdir is empty, it will lead to new collision since it would
+	// succeed but point at the cache's root
+	if subdir == "" {
+		return "", fmt.Errorf("invalid parameter")
+	}
 
-	absdir, err := filepath.Abs(filepath.Join(root, subdir))
+	absdir, err := filepath.Abs(filepath.Join(c.Root, subdir))
 	if err != nil {
-		sylog.Fatalf("Unable to get abs filepath: %v", err)
+		return "", fmt.Errorf("unable to get abs filepath: %v", err)
 	}
 
 	if err := initCacheDir(absdir); err != nil {
-		sylog.Fatalf("Unable to initialize caching directory: %v", err)
+		return "", fmt.Errorf("unable to initialize caching directory: %v", err)
 	}
 
 	sylog.Debugf("Caching directory set to %s", absdir)
-	return absdir
+
+	return absdir, nil
 }
 
+// initCacheDir initializes a sub-cache within a cache, e.g., the shub sub-cache.
 func initCacheDir(dir string) error {
 	if _, err := os.Stat(dir); os.IsNotExist(err) {
```

Comment:
since this is already paying the price of calling stat, it should use the returned os.FileInfo to ask if `dir` is in fact a directory (`fi.IsDir()`)

Reply:
I was checking if it is a directory right after but can change it to use the FileInfo for sure. Good catch.

Reply:
Done.